packages/pulseaudio/module-sles-sink.c

```diff
@@ -125,23 +130,197 @@ static void sles_callback(SLBufferQueueItf bqPlayerBufferQueue, void *userdata)
+                pa_log("Too many channels in sl mask");
+                return -1;
+            }
+        }
+    }
+    if (mask) {
+        pa_log("Unknown channel in sl mask (left: %u)", mask);
+        return -1;
+    }
+    *rmap = map;
+    return 0;
+}
+
+// pa_channel_map_to_sl_channel_mask converts a PulseAudio channel map to an
+// OpenSL ES channel mask. If an unknown or unsupported channel position is
+// found, -1 is returned. If map is NULL, 0 is returned.
+//
+// https://freedesktop.org/software/pulseaudio/doxygen/channelmap_8h.html#af1cbe2738487c74f99e613779bd34bf2
+// https://www.khronos.org/files/opensl_es_1_0_provisional_specs.pdf#9.1.8
+static SLuint32 pa_channel_map_to_sl_channel_mask(pa_channel_map *map) {
+    SLuint32 mask = 0, cur = 0, last = 0;
+    if (!map)
+        return 0;
+    for (int i = 0; i < map->channels; i++) {
+        switch (map->map[i]) {
+            case PA_CHANNEL_POSITION_MONO:                  cur = SL_SPEAKER_FRONT_CENTER;          break;
+            case PA_CHANNEL_POSITION_FRONT_LEFT:            cur = SL_SPEAKER_FRONT_LEFT;            break;
+            case PA_CHANNEL_POSITION_FRONT_RIGHT:           cur = SL_SPEAKER_FRONT_RIGHT;           break;
+            case PA_CHANNEL_POSITION_FRONT_CENTER:          cur = SL_SPEAKER_FRONT_CENTER;          break;
+            case PA_CHANNEL_POSITION_REAR_CENTER:           cur = SL_SPEAKER_BACK_CENTER;           break;
+            case PA_CHANNEL_POSITION_REAR_LEFT:             cur = SL_SPEAKER_BACK_LEFT;             break;
+            case PA_CHANNEL_POSITION_REAR_RIGHT:            cur = SL_SPEAKER_BACK_RIGHT;            break;
+            case PA_CHANNEL_POSITION_LFE:                   cur = SL_SPEAKER_LOW_FREQUENCY;         break;
+            case PA_CHANNEL_POSITION_FRONT_LEFT_OF_CENTER:  cur = SL_SPEAKER_FRONT_LEFT_OF_CENTER;  break;
+            case PA_CHANNEL_POSITION_FRONT_RIGHT_OF_CENTER: cur = SL_SPEAKER_FRONT_RIGHT_OF_CENTER; break;
+            case PA_CHANNEL_POSITION_SIDE_LEFT:             cur = SL_SPEAKER_SIDE_LEFT;             break;
+            case PA_CHANNEL_POSITION_SIDE_RIGHT:            cur = SL_SPEAKER_SIDE_RIGHT;            break;
+            case PA_CHANNEL_POSITION_TOP_CENTER:            cur = SL_SPEAKER_TOP_CENTER;            break;
+            case PA_CHANNEL_POSITION_TOP_FRONT_LEFT:        cur = SL_SPEAKER_TOP_FRONT_LEFT;        break;
+            case PA_CHANNEL_POSITION_TOP_FRONT_RIGHT:       cur = SL_SPEAKER_TOP_FRONT_RIGHT;       break;
+            case PA_CHANNEL_POSITION_TOP_FRONT_CENTER:      cur = SL_SPEAKER_TOP_FRONT_CENTER;      break;
+            case PA_CHANNEL_POSITION_TOP_REAR_LEFT:         cur = SL_SPEAKER_TOP_BACK_LEFT;         break;
+            case PA_CHANNEL_POSITION_TOP_REAR_RIGHT:        cur = SL_SPEAKER_TOP_BACK_RIGHT;        break;
+            case PA_CHANNEL_POSITION_TOP_REAR_CENTER:       cur = SL_SPEAKER_TOP_BACK_CENTER;       break;
```

Comment:
Or perhaps even just an array of SL_SPEAKER_* with PA_CHANNEL_POSITION as the indices...

Reply:
I took another look at this, and it seems that I forgot to replace the switch here with a function call to `pa_channel_position_to_sl_speaker` when extracting that code.

Reply:
> PA_CHANNEL_POSITION_* as the indices

I suppose designated inits would work fine since unlike the SL constants, the PA ones have a reasonable max (they're enums rather than bitmasks). I'll implement your suggestion.

Reply:
Done.